### PR TITLE
Migrate multi-currency e2e tests to Playwright

### LINKED ISSUE #8750: [e2e] Migrate multi-currency e2e tests to playwright
## Description

In an effort to improve the maintainability of the multi-currency e2e tests we want to migrate them to Playwright. This should offer more reliability when running the tests.

The following tests will have to be migrated:

- `tests/e2e/specs/wcpay/merchant/merchant-admin-multi-currency.spec.js`
- `tests/e2e/specs/wcpay/merchant/merchant-admin-multi-currency-on-boarding.spec.js`
- `tests/e2e/specs/wcpay/merchant/merchant-admin-multi-currency-setup.spec.js`
- `tests/e2e/specs/wcpay/shopper/shopper-checkout-multi-currency.spec.js`

It _may_ be worth it to split this into 2 issues (merchant/shopper) when we start working on this.
```[tasklist]
### Tasks
- [x] `tests/e2e/specs/wcpay/merchant/merchant-admin-multi-currency.spec.js`
- [x] `tests/e2e/specs/wcpay/merchant/merchant-admin-multi-currency-on-boarding.spec.js`
- [x] `tests/e2e/specs/wcpay/merchant/merchant-admin-multi-currency-setup.spec.js`
- [x] `tests/e2e/specs/wcpay/shopper/shopper-checkout-multi-currency.spec.js`
```

### PR DESCRIPTION
**Pull request by cesarcosta99**
Resolves #8750.

#### Changes proposed in this Pull Request

This PR migrates all multi-currency specs we have in Puppeteer to Playwright:

- `merchant-admin-multi-currency.spec.js` → `multi-currency.spec.ts`
- `merchant-admin-multi-currency-on-boarding.spec.js` → `multi-currency-on-boarding.spec.ts`
- `merchant-admin-multi-currency-setup.spec.js` → `multi-currency-setup.spec.ts`
- `shopper-checkout-multi-currency.spec.js` → `multi-currency-checkout.spec.ts`

Some background on decisions I've made while migrating the specs:

- The Playwright setup and new specs has been written in TypeScript, so I wrote the new helpers and multi-currency specs in TypeScript too.
- I dropped the `merchant-admin/shopper` prefix in the filenames because they are within folders with the same name and it sounds redundant.
- I kept the old utility functions in Puppeteer to reduce risks and also because they might be needed while we don't fully migrate.

#### Testing instructions

**Setup**

1. Make sure you have the e2e environment setup before running the tests: `npm run test:e2e-setup`.
2. If you need to set it up from scratch, see [the README](https://github.com/Automattic/woocommerce-payments/blob/d89175a31643899e3abd8305e39ccefef899a5fd/tests/e2e/README.md).

**Ensure all GH actions below are passing, particularly the E2E Playwright Tests**.

**Test: Ensure all tests pass locally consistently**

You can run the e2e tests multiple times and try different combinations to confirm they all pass and are stable. Some suggestions:

- `npm run test:e2e-pw multi-currency`
- `npm run test:e2e-pw shopper/multi-currency`
- `npm run test:e2e-pw merchant/multi-currency`
- `npm run test:e2e-pw <multi-currency-spec-name>`

**Test: Ensure the migrated specs test what they're expected to test**

You can ensure that by either comparing the old spec code with the new one, and/or watching the tests running via the UI command: `npm run test:e2e-pw-ui multi-currency`.

-------------------

- [x] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [x] Covered with tests (or have a good reason not to test in description ☝️)
- [x] Tested on mobile (or does not apply)

**Post merge**

- [ ] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) following [these instructions](https://github.com/Automattic/woocommerce-payments/wiki/How-to-write-good-manual-testing-scenarios) : _Add link here / 'QA Testing Not Applicable'_
- [ ] Add or update [critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Critical-flows) and [testing instructions for critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Testing-instructions-for-critical-flows), if applicable.
- [ ] Add what's changed (description, screenshot, demo videos etc.) to the release announcement post, if applicable.

### CODE REVIEW

**Comment by botwoo** [comment]
### Test the build 

#### Option 1. Jetpack Beta

- Install and activate [Jetpack Beta](https://jetpack.com/download-jetpack-beta/).
- Use this build by searching for PR number `9175` or branch name `dev/8750-migrate-mccy-e2e-tests-to-playwright` in your-test.site/wp-admin/admin.php?page=jetpack-beta&plugin=woocommerce-payments          

#### Option 2. Jurassic Ninja - available for logged-in A12s

:rocket: [Launch a JN site with this branch](https://jurassic.ninja/create/?jetpack-beta&shortlived&nojetpack&woocommerce&woocommerce-payments-dev-tools&branches.woocommerce-payments=dev/8750-migrate-mccy-e2e-tests-to-playwright) :rocket:

:information_source: Install this [Tampermonkey script](https://github.com/Automattic/woocommerce-payments/tree/develop/bin/wcpay-live-branches) to get more options.

---

Build info: 

- Latest commit: 2fed4d89429b38c8c7ceadf3875fbc7ecd37b134
- Build time: 2024-08-20 16:47:57 UTC

_Note: the build is updated when a new commit is pushed to this PR._

File: tests/e2e-pw/utils/shopper.ts [hunk 1]
```diff
@@ -122,3 +125,131 @@ export const confirmCardAuthentication = async (
...
+
+/**
+ * Retrieves the product price from the current product page.
+ *
+ * This function assumes that the page object has already navigated to a product page.
+ */
+export const getPriceFromProduct = async (
+	page: Page,
+	slug: string
+): Promise< string > => {
```
Comment (reykjalin) [inline tests/e2e-pw/utils/shopper.ts]: > [!NOTE]
> This is a nitpick.

Generally we recommend not adding return type declarations anywhere and instead leave it to TS to infer. This function is simple enough, and the return type simple enough, that I think this is fine, but something to keep in mind going forward :)

The same return type declaration happens in the last function in this file.

File: tests/e2e-pw/specs/shopper/multi-currency-checkout.spec.ts [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+		await addCurrency( merchantPage, 'EUR' );
+	} );
+
+	test.afterAll( async () => {
+		await removeCurrency( merchantPage, 'EUR' );
+
+		if ( ! wasMulticurrencyEnabled ) {
+			await deactivateMulticurrency( merchantPage );
+		}
+	} );
```
Comment (reykjalin) [inline tests/e2e-pw/specs/shopper/multi-currency-checkout.spec.ts]: I notice the previous test had various `merchant.login()` and `shopper.login()` calls with corresponding `merchant.logout()` and `shopper.logout()` calls in the `beforeAll` and `afterAll` functions. I wonder if that's required here too?

Is that what `shopperPage` and `merchantPage` do? Create separate sessions with independent navigation and browser state?

File: tests/e2e-pw/specs/merchant/multi-currency-setup.spec.ts [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+		shopperPage = ( await getShopper( browser ) ).shopperPage;
+		merchantPage = ( await getMerchant( browser ) ).merchantPage;
+		wasMulticurrencyEnabled = await activateMulticurrency( merchantPage );
+	} );
+
+	test.afterAll( async () => {
+		if ( ! wasMulticurrencyEnabled ) {
+			await deactivateMulticurrency( merchantPage );
+		}
+	} );
```
Comment (reykjalin) [inline tests/e2e-pw/specs/merchant/multi-currency-setup.spec.ts:37]: I notice the previous test had various `merchant.login()` calls with corresponding `merchant.logout()` calls in the `beforeAll` and `afterAll` functions. I wonder if that's required here too?

Is that what `shopperPage` and `merchantPage` do? Create separate sessions with independent navigation and browser state?

**Review by reykjalin — CHANGES_REQUESTED** [review]
Love to see the tests being migrated like this and it's an excellent indicator that the tests are passing in GH actions!

However, I ran `npm run test:e2e-pw multi-currency` 4 times and it failed 3 out of the 4 times. The zip file of the results is in the team channel in Slack. 2 of the failures were identical, the last failure only failed one of the tests, but in the same way as the first 2 failures:

```
  4 failed
    [merchant] › merchant/multi-currency.spec.ts:38:6 › Multi-currency › page load without any errors
    [shopper] › shopper/multi-currency-checkout.spec.ts:45:8 › Multi-currency checkout › Checkout with multiple currencies › checkout with EUR
    [shopper] › shopper/multi-currency-checkout.spec.ts:52:8 › Multi-currency checkout › Checkout with multiple currencies › should display EUR in the order received page
    [shopper] › shopper/multi-currency-checkout.spec.ts:60:8 › Multi-currency checkout › Checkout with multiple currencies › should display EUR in the customer order page
```

The single failure:

```
  1 failed
    [merchant] › merchant/multi-currency.spec.ts:38:6 › Multi-currency › page load without any errors
```

I got similar results when I ran `npm run test:e2e-pw merchant/multi-currency` and `npm run test:e2e-pw shopper/multi-currency`; same tests failed in the same places when they did not succeed. They failed about half the time to 2/3rds of the time.

Since it's the same tests failing every time it probably has something to do with the tests themselves rather than anything else?

It's also possible this has something to do with my environment. I just set up the e2e testing for this PR so maybe I did something wrong, or didn't do something correctly 🤔

File: tests/e2e-pw/specs/shopper/multi-currency-checkout.spec.ts (hunk 2 above)
Comment (cesarcosta99) [reply tests/e2e-pw/specs/shopper/multi-currency-checkout.spec.ts]: >Is that what shopperPage and merchantPage do? Create separate sessions with independent navigation and browser state?

Precisely. That's documented in the [FAQ](https://github.com/Automattic/woocommerce-payments/blob/develop/tests/e2e-pw/README.md#faqs) and we don't need to login/logout between shopper and merchant anymore 🎉

File: tests/e2e-pw/specs/merchant/multi-currency-setup.spec.ts (hunk 3 above)
Comment (cesarcosta99) [reply tests/e2e-pw/specs/merchant/multi-currency-setup.spec.ts:37]: Answered that in the [previous comment](https://github.com/Automattic/woocommerce-payments/pull/9175#discussion_r1719063589).

**Comment by cesarcosta99** [comment]
@reykjalin, as I mentioned on Slack, I wasn't able to reproduce the errors from your report, instead I was consistently running into a different error: the order received one.

I've made some changes to make the tests more stable. I've run `npm run test:e2e-pw merchant/multi-currency` and `npm run test:e2e-pw shopper/multi-currency` locally 10 times each and all tests passed on all runs. Can you please pull and retry? If you continue to experience errors we can pair tomorrow and try to identify what's wrong.

File: tests/e2e-pw/utils/shopper.ts (hunk 1 above)
Comment (cesarcosta99) [reply tests/e2e-pw/utils/shopper.ts]: I didn't know that was a recommendation but I agree. Changed it in [2083568](https://github.com/Automattic/woocommerce-payments/pull/9175/commits/2083568acc7f1c8d7e20fd13e8c159d937918a21).

**Comment by cesarcosta99** [comment]
Just for reference, this is the Slack thread we have been discussing the latest changes: p1723693966218459-slack-C01BZUL57SQ

**Review by ricardo — COMMENTED** [review]
@cesarcosta99 Running the tests locally, the only test that fails consistently for me is `should display the correct currency in the my account order history table` at `shopper/multi-currency-checkout.spec.ts:84:7`:

```
Locator: locator('tr').filter({ has: getByLabel('View order number 142') })
Expected pattern: /USD/
Received string:  ""
```

The test appears to pass in the CI:
```
1 skipped
34 passed
```

**Review by ricardo — APPROVED** [review]
```
Locator: locator('tr').filter({ has: getByLabel('View order number 142') })
Expected pattern: /USD/
Received string:  ""
```

I figured why I was experiencing the error above. I had installed the test environment from scratch but I guess I hadn't removed the cached dependencies, so my WooCommerce version was `8.8.3` still.

Updating it made the test work as the selector was introduced somewhere in between in WooCommerce.

All tests are working consistently for me now. ✅

**Review by reykjalin — APPROVED** [review]
I think the remaining problem are just my environment being slow (see aforementioned slack thread) so I'm approving this. The tests look good and generally run well 🙌 

Awesome work! 🚀

**Pull request merged**